api/datastore/mysql/mysql.go

```diff
@@ -0,0 +1,615 @@
+	err := ds.Tx(func(tx *sql.Tx) error {
+		row := ds.db.QueryRow(fmt.Sprintf("%s WHERE app_name=? AND path=?", routeSelector), newroute.AppName, newroute.Path)
+		if err := scanRoute(row, &route); err == sql.ErrNoRows {
+			return models.ErrRoutesNotFound
+		} else if err != nil {
+			return err
+		}
+
+		route.Update(newroute)
+
+		hbyte, err := json.Marshal(route.Headers)
+		if err != nil {
+			return err
+		}
+
+		cbyte, err := json.Marshal(route.Config)
+		if err != nil {
+			return err
+		}
+
+		res, err := tx.Exec(`
+		UPDATE routes SET
+			image = ?,
+			format = ?,
+			maxc = ?,
+			memory = ?,
+			type = ?,
+			timeout = ?,
+			headers = ?,
+			config = ?
+		WHERE app_name = ? AND path = ?;`,
+			route.Image,
+			route.Format,
+			route.MaxConcurrency,
+			route.Memory,
+			route.Type,
+			route.Timeout,
+			string(hbyte),
+			string(cbyte),
+			route.AppName,
+			route.Path,
+		)
+
+		if err != nil {
+			return err
+		}
+
+		if n, err := res.RowsAffected(); err != nil {
+			return err
+		} else if n == 0 {
+			return models.ErrRoutesNotFound
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	return &route, nil
+}
+
+/*
+RemoveRoute ...
+*/
+func (ds *MySQLDatastore) RemoveRoute(ctx context.Context, appName, routePath string) error {
+	res, err := ds.db.Exec(`
+		DELETE FROM routes
+		WHERE path = ? AND app_name = ?
+	`, routePath, appName)
+
+	if err != nil {
+		return err
+	}
+
+	n, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+
+	if n == 0 {
+		return models.ErrRoutesRemoving
+	}
+
+	return nil
+}
+
+func scanRoute(scanner rowScanner, route *models.Route) error {
+	var headerStr string
+	var configStr string
+
+	err := scanner.Scan(
+		&route.AppName,
+		&route.Path,
+		&route.Image,
+		&route.Format,
+		&route.MaxConcurrency,
+		&route.Memory,
+		&route.Type,
+		&route.Timeout,
+		&headerStr,
+		&configStr,
+	)
+
+	if headerStr == "" {
+		return models.ErrRoutesNotFound
+	}
+
+	json.Unmarshal([]byte(headerStr), &route.Headers)
+	json.Unmarshal([]byte(configStr), &route.Config)
+
+	return err
+}
+
+/*
+GetRoute ...
+*/
+func (ds *MySQLDatastore) GetRoute(ctx context.Context, appName, routePath string) (*models.Route, error) {
+	var route models.Route
+
+	row := ds.db.QueryRow(fmt.Sprintf("%s WHERE app_name=? AND path=?", routeSelector), appName, routePath)
+	err := scanRoute(row, &route)
+
+	if err == sql.ErrNoRows {
+		return nil, models.ErrRoutesNotFound
+	} else if err != nil {
+		return nil, err
+	}
+	return &route, nil
+}
+
+/*
+GetRoutes ...
+*/
+func (ds *MySQLDatastore) GetRoutes(ctx context.Context, filter *models.RouteFilter) ([]*models.Route, error) {
+	res := []*models.Route{}
+	filterQuery, args := buildFilterRouteQuery(filter)
+	rows, err := ds.db.Query(fmt.Sprintf("%s %s", routeSelector, filterQuery), args...)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var route models.Route
+		err := scanRoute(rows, &route)
+		if err != nil {
+			continue
```

Comment:
I mentioned this elsewhere, but it's worth pointing out in context: `scanRoute()` errors are tossed here.  Although it might make sense to return a best effort to the users, the error is probably worth logging since it's likely bad data or buggy scanning.